### PR TITLE
RFC: WIP: add support for compiler directives which apply to functions

### PR DESCRIPTION
**Pull request by tblah**
RFC: not for merge

Some directives (e.g. `noinline`) apply to the following function-like statement. This PR provides partial support for these kinds of directives, using some new directives for Arm streaming SVE (modeled after the clang attributes with the same name).

The directives are currently supported for function-like statements at the top-level or inside of modules.

I would like comments on the rough shape of the implementation (especially the changes to `PFTBuilder.cpp` and the decision to match directives to functions in `Bridge.cpp`).

## To Do
- [ ]  We currently cannot parse compiler directives inside of interface blocks
- [ ]  No support for nested function-like statements
- [ ]  No support for `.mod` files
- [ ]  The warning generated when a directive is not followed by a function-like statement is not using proper diagnostics and prints the source location inside of `loc<>`

### CODE REVIEW

File: flang/lib/Semantics/resolve-names.cpp [hunk 1]
```diff
@@ -8389,7 +8389,31 @@ void ResolveNamesVisitor::Post(const parser::CompilerDirective &x) {
       }
     }
   } else {
-    Say(x.source, "Compiler directive was ignored"_warn_en_US);
+    bool handled = false;
```
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:8392]: `{false};`

File: flang/lib/Semantics/resolve-names.cpp [hunk 2]
```diff
@@ -8389,7 +8389,31 @@ void ResolveNamesVisitor::Post(const parser::CompilerDirective &x) {
...
+    if (const auto *nvList{
+            std::get_if<std::list<parser::CompilerDirective::NameValue>>(
+                &x.u)}) {
+      for (const parser::CompilerDirective::NameValue &nv : *nvList) {
+        std::string name = std::get<parser::Name>(nv.t).ToString();
```
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:8397]: Braced initialization please everywhere in semantics

**Comment by tblah** [comment]
Thanks for taking a look @klausler. I've fixed the comments.

Does the overall approach look acceptable?

**Comment by klausler** [comment]
> Thanks for taking a look @klausler. I've fixed the comments.
> 
> Does the overall approach look acceptable?

You bet, looks great.

**Comment by tblah** [comment]
There is a much more complete version of this here https://github.com/llvm/llvm-project/pull/95123

**Pull request closed without merging**